### PR TITLE
Set up a php oc10 controller for reading the config.json from the config folder

### PR DESCRIPTION
**Pull request by kulmann**
## Description
When ownCloud Web is deployed as an app, we need to read the config from the config folder, instead of from within the app folder. Otherwise we'll have issues with the signing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by kulmann** [comment]
FYI, I discovered that while the app is reachable from a route with `apps-external` in it, the controller endpoint is only reachable from `<baseurl>/index.php/apps/web/config`, not `<baseurl>/index.php/apps-external/web/config`.

**Review by LukasHirt — APPROVED** [review]
I actually liked it when the config was directly copied as `config.json`. No need to rename/copy the original file then on the server. But it's not anything critical that I'd try to block the PR with. :grin: LGTM 🚀

**Comment by kulmann** [comment]
> I actually liked it when the config was directly copied as `config.json`. No need to rename/copy the original file then on the server. But it's not anything critical that I'd try to block the PR with. 😁 LGTM 🚀

The reason why this is important in this PR is because we first try to load the config.json directly and only if it doesn't exist we try the new oc10 app endpoint. So we really need that it doesn't exist in the first place. 😁 

One ugly thing here: for the app deployment we will always have a 404 error for the `config.json` in the javascript console. The 404 is expected and we react on it by trying the config endpoint. But we can't suppress the console output.

**Comment by micbar** [comment]
@C0rby This needs a security review. Can you take a look?

**Comment by kulmann** [comment]
> > I actually liked it when the config was directly copied as `config.json`. No need to rename/copy the original file then on the server. But it's not anything critical that I'd try to block the PR with. 😁 LGTM 🚀
> 
> The reason why this is important in this PR is because we first try to load the config.json directly and only if it doesn't exist we try the new oc10 app endpoint. So we really need that it doesn't exist in the first place. 😁
> 
> One ugly thing here: for the app deployment we will always have a 404 error for the `config.json` in the javascript console. The 404 is expected and we react on it by trying the config endpoint. But we can't suppress the console output.

This is not true anymore. Instead of relying on the webserver to serve the app, we now have a php controller in place that serves all the files (except for the config.json) of the app. This brings three improvements:
1. We were now able to change the route for the config, so that it's `config.json` again and accessed with a `require('config.json')` call again. No need to construct the php controller route in the js code. No knowledge of the php env in the js code anymore. 🎉  🙌 
2. For configuring the `web.baseUrl` in config.php, the admin doesn't need to care anymore whether the `web` app lives in `apps`, `apps-external` or some other custom apps folder. The Url always points to `apps`, as the app framework within core takes care of merging all configured app locations into `apps` routes.
3. When the app is disabled, it (properly) doesn't work anymore. Before it was still served from the webserver, so disabling the app didn't have any effect. The only remaining thing is the nav item. So after disabling the app, one still has to remove the `web.baseUrl` from config.php to remove all traces of the app in the frontend.

**Comment by kulmann** [comment]
> @C0rby This needs a security review. Can you take a look?

@C0rby 
Some context and more detailed questions: we introduced a `FilesController` which takes care of serving all the files of the web app (js, css, html, ...). Within the controller we have some checks in place - but we would like to have a second opinion of whether this is sufficient:
1. we don't allow `..` in any path so that directory traversals to parent folders outside the app should ™️ not be possible.
2. we only allow paths and files that start with or equal the content of the `permittedPaths` array.
3. we needed to allow inline script execution in CSP headers for the two files  `oidc-callback.html` and `oidc-silent-redirect.html`, because they are redirecting to different domains via inline script execution. We ran some manual tests if those headers really only apply to those two files - looks like that is the case. But this needs your attention and opinion.
4. since we're serving the files from the controller, the rules from the `.htaccess` file don't apply. We applied the headers manually the the response headers.

cc @LukasHirt

**Review by micbar — CHANGES_REQUESTED** [review]
Let us wait for @C0rby before we merge

**Comment by C0rby** [comment]
The code looks good so far but I wanted to try to bypass the path traversal checks.
Unfortunately I can't get it to run... I always get `config.json not found` but it doesn't seem like the `ConfigController.php` is called.

**Comment by kulmann** [comment]
> The code looks good so far but I wanted to try to bypass the path traversal checks.
> Unfortunately I can't get it to run... I always get `config.json not found` but it doesn't seem like the `ConfigController.php` is called.

Did you copy a `config.json` into the `config` path of oc10?

**Comment by C0rby** [comment]
Yes, but my error was that I forgot the `index.php` in the baseUrl... :facepalm:

File: lib/Controller/FilesController.php
Comment (C0rby) [inline lib/Controller/FilesController.php:90]: The checks look good.
There is one thing I would change though.
```php
$basePath = \dirname(__DIR__,2);
$absolutePath = \realpath( $basePath . '/' . $path;
if ($absolutePath === false) {
  return new DataResponse(['error' => 'resource not found'], Http::STATUS_NOT_FOUND);
}
if (\strpos($absolutePath, $basePath) !== 0) {
  return new DataResponse(['error' => 'resource not found'], Http::STATUS_NOT_FOUND);
}
```

This way if somehow an attacker manages to do path traversal we still check if the resulting path is in the allowed directory.

**Review by C0rby — APPROVED** [review]
This looks good. :+1: 
You might want to consider what I wrote in the comment but it's not absolutely necessary.

**Review by micbar — APPROVED** [review]

**Pull request merged**
